querybook/server/lib/query_analysis/validation/validators/presto_optimizing_validator.py

```diff
@@ -171,53 +214,117 @@ def get_query_validation_results(
 
-    def _get_presto_explain_validation_results(
-        self, query: str, uid: int, engine_id: int
-    ) -> List[QueryValidationResult]:
-        return self._get_explain_validator().validate(query, uid, engine_id)
+    def _search_columns_for_suggestion(self, columns: List[str], suggestion: str):
+        """Return the case-sensitive column name by searching the table's columns for the suggestion text"""
+        for col in columns:
+            if col.lower() == suggestion.lower():
+                return col
+        return suggestion
+
+    def _get_column_name_suggestion(
+        self,
+        validation_result: QueryValidationResult,
+        query: str,
+        tables_in_query: List[str],
+        raw_tokens: List[Token] = None,
+    ):
+        fuzzy_column_name = self._get_column_name_from_position(
+            query,
+            validation_result.start_line,
+            validation_result.start_ch,
+            raw_tokens=raw_tokens,
+        )
+        if not fuzzy_column_name:
+            return None
+        results, count = search_table.get_column_name_suggestion(
+            fuzzy_column_name, tables_in_query
+        )
+        if count == 1:  # Only return suggestion if there's a single match
+            table_result = results[0]
+            highlights = table_result.get("highlight", {}).get("columns", [])
+            if len(highlights) == 1:
+                column_suggestion = self._search_columns_for_suggestion(
+                    table_result.get("columns"), highlights[0]
+                )
+                return column_suggestion
+
+        return None
 
     def validate(
         self,
         query: str,
         uid: int,
         engine_id: int,
+        raw_tokens: List[QueryValidationResult] = None,
+        **kwargs,
     ) -> List[QueryValidationResult]:
-        validation_results = [
-            *self._get_presto_explain_validation_results(query, uid, engine_id),
-            *self._get_sql_glot_validation_results(query),
-        ]
+        if raw_tokens is None:
+            raw_tokens = self._tokenize_query(query)
+        validation_results = self._validator.validate(
+            query, uid, engine_id, raw_tokens=raw_tokens
+        )
+        tables_in_query = self._get_tables_in_query(query, engine_id)
+        for result in validation_results:
+            if self._is_column_name_error(result):
+                column_suggestion = self._get_column_name_suggestion(
+                    result, query, tables_in_query, raw_tokens
+                )
+                if column_suggestion:
+                    result.suggestion = column_suggestion
         return validation_results
+
+
+class PrestoOptimizingValidator(BaseQueryValidator):
+    def languages(self):
+        return ["presto", "trino"]
+
+    def _get_explain_validator(self):
+        return PrestoExplainValidator("")
+
+    def _get_decorated_validator(self) -> BaseQueryValidator:
+        return UnionAllValidator(
```

Comment:
what's the reason of changing from a list of validators to a chain of validators?

Reply:
Using the decorator pattern so that we can add suggestions on top of the validation messages